mantidproject/mantid
# PyChop update

**Description of work.**



Adds handling for arbitrary number of independently phased choppers (issue #29965). This is mostly a GUI issue as the calculation code should have been able to handle this already (although there was a small issue here in parsing the input `yaml` file as it was assuming only a single chopper would be phased independently of the source).

Also some code clean up:

* Add GUI test for the independent phases.
* Remove `raise` exceptions which are not explicitly caught in the code and replace them with error messages
* Refactor some of the code for special handling for Merlin and added a test for this
* Add GUI test for the "enter an Ei value to automatically replot" mode.

**To test:**

Checks that the test passes.

Open PyChop with this instrument (a modification of CNCS):
[cncs_indep.zip](https://github.com/mantidproject/mantid/files/7591307/cncs_indep.zip)

And check that there are labels and edit boxes for "Chopper 1 Phase delay time" and "Chopper 3 Phase delay time" compared with normal CNCS.



Fixes #29965 and #30716 . 







---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
